R/XcmsExperiment-functions.R

```diff
@@ -878,40 +883,49 @@
     fd <- fData(chrs)
     rtc <- c("rtmin", "rtmax")
     mzc <- c("mzmin", "mzmax")
-    cpd <- chromPeakData(object)
-    pks_empty <- chromPeaks(object)[integer(), ]
-    pkd_empty <- cpd[integer(), ]
+    if (inherits(object, "XcmsExperiment"))
+        cpd <- .chromPeakData(object)
+    else cpd <- as.data.frame(chromPeakData(object))
+    message("Processing chromatographic peaks")
+    pb <- progress_bar$new(format = paste0("[:bar] :current/:",
+                                           "total (:percent) in ",
+                                           ":elapsed"),
+                           total = nrow(chrs) + 1L, clear = FALSE)
     for (i in seq_len(nrow(chrs))) {
-        pks <- chromPeaks(object, rt = fd[i, rtc], mz = fd[i, mzc],
-                          msLevel = chrs[i, 1]@msLevel, type = chromPeaks)
+        idx <- .index_chrom_peaks(
+            object, rt = fd[i, rtc], mz = fd[i, mzc],
+            msLevel = chrs[i, 1]@msLevel, type = chromPeaks)
+        f_s <- factor(.chromPeaks(object)[idx, "sample"], levels = js)
+        pkl <- split.data.frame(.chromPeaks(object)[idx, , drop = FALSE], f_s)
```

Comment:
splitting the chrom peaks `matrix` and chrom peak data `data.frame` considerably improved performance. Also, we try to use `data.frame` instead of `DataFrame` for `chromPeakData` as much as possible: subset of the former is much faster.

Reply:
Can I ask the reason of why the `DataFrame` was initially used ? was it easier at the beginning ? I have not really end up using `DataFrame` until now so I'm a bit curious as to when someone decide to prefer this over `data.frame`

Reply:
Good question. There were 2 reasons: firstly, the `DataFrame` was used in e.g. `MSnbase` and other Bioconductor packages (e.g. `SummarizedExperiment`) and I wanted to be consistent with that, then, secondly, the `DataFrame` would also allow to store S4 objects in columns (which the `data.frame` would *officially* not). Not that we planned to store now S4 objects in `DataFrame`s - so it was mostly reason 1 why we used `DataFrame`.

Only (quite some time) later I figured out that subset or extract data from a `DataFrame`s is quite slower than for `data.frame`. Thus, recently, I prefer to use, whenever possible (internally) a `data.frame`.